### ISSUE TITLE
Make core model interfaces return docblocks more specific

### ISSUE THREAD

**Issue by lasotaartur** [Sylius/Sylius#8224]
Hi,  I have problem everywhere with sylius interface i.e.

```
$product = $productVariant->getProduct();
```
will return object with interface `Sylius\Component\Product\Model\ProductInterface`

then when I write this
```
$taxon = $product->getMainTaxon();
```
PHPstorm will show that the not exists in interface `Sylius\Component\Product\Model\ProductInterface` but exists in `\Sylius\Component\Core\Model\ProductInterface` (edited)

And problem exists for others components also

**Comment by lchrusciel** [comment]
Yes, we are aware of this. That way, there is a lot of variable type hints around Sylius, But this probably won't be resolved, as it is a trade-off, that we are probably willing to pay. Unless somebody has an idea how to make it better. Not sure, if there is any point in keeping it open? /cc @pamil @pjedrzejewski

**Comment by pamil** [comment]
Actually, we might end up overriding return type docblocks in Core entities so that it won't happen if traversing from another Core entity.

**Comment by pjedrzejewski** [comment]
This is one of the most annoying DX issues we have IMO, let's try to fix it for 1.0.0.

**Comment by pamil** [comment]
Tried to implement it before beta.3, but return types will make it much more easier.
